lib/web/recordingplayback_test.go

```diff
@@ -0,0 +1,723 @@
+
+	require.Equal(t, eventTypeBatch, responses[2][0], "Third message should be batch event")
+
+	require.Contains(t, string(responses[2][responseHeaderSize:]), "f", "Second batch should contain event with f data")
+	require.Contains(t, string(responses[2][responseHeaderSize:]), "g", "Second batch should contain event with g data")
+
+	require.Equal(t, eventTypeStop, responses[3][0], "Fourth message should be stop event")
+}
+
+func createFetchRequest(start, end int64, requestID uint32, currentScreen bool) []byte {
+	buf := make([]byte, requestHeaderSize)
+
+	buf[0] = requestTypeFetch
+
+	binary.BigEndian.PutUint64(buf[1:9], uint64(start))
+	binary.BigEndian.PutUint64(buf[9:17], uint64(end))
+	binary.BigEndian.PutUint32(buf[17:21], requestID)
+
+	buf[21] = 0
+	if currentScreen {
+		buf[21] = 1
+	}
+
+	return buf
+}
+
+// fetchAndCollectResponses sends a fetch request over the WebSocket connection and collects all responses.
+func fetchAndCollectResponses(t *testing.T, ws *websocket.Conn, start, end int64, requestCurrentScreen bool) [][]byte {
+	testTimeout := 5 * time.Second
+
+	req := createFetchRequest(start, end, 1, requestCurrentScreen)
+
+	err := ws.WriteMessage(websocket.BinaryMessage, req)
+	require.NoError(t, err)
+
+	var responses [][]byte
+
+	done := make(chan bool)
+	go func() {
+		for {
+			ws.SetReadDeadline(time.Now().Add(testTimeout))
+			_, msg, err := ws.ReadMessage()
+			if err != nil {
+				break
+			}
+
+			responses = append(responses, msg)
+			if len(msg) > 0 && msg[0] == eventTypeStop {
+				done <- true
+				break
+			}
+		}
+	}()
+
+	select {
+	case <-done:
+		break
+	case <-time.After(testTimeout):
+		t.Fatal("Timeout waiting for responses")
+	}
+
+	return responses
+}
+
+// createWebSocket sets up a WebSocket server for testing, returning the server, websocket connection and mock
+// client, taking a callback to allow populating the mock client with events before running the playback.
+func createWebSocket(t *testing.T, setupEvents func(mockClient *mockStreamClient)) (*httptest.Server, *websocket.Conn, *mockStreamClient) {
+	mockClient := newMockStreamClient()
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		upgrader := websocket.Upgrader{}
+		ws, err := upgrader.Upgrade(w, r, nil)
+		require.NoError(t, err)
+		defer ws.Close()
+
+		ctx := context.Background()
+		logger := slog.Default()
+
+		playback := newRecordingPlayback(ctx, ws, mockClient, "test-session", logger)
+
+		go setupEvents(mockClient)
+
+		playback.run()
+	}))
+
+	wsURL := "ws" + strings.TrimPrefix(server.URL, "http")
+	ws, _, err := websocket.DefaultDialer.Dial(wsURL, nil)
+	require.NoError(t, err)
+
+	return server, ws, mockClient
+}
+
+// closeWebSocket gracefully closes the WebSocket connection by sending a close message.
+func closeWebSocket(t *testing.T, ws *websocket.Conn) {
```

Comment:
Should we name it `sendCloseMessage` instead? I would have expected to see a `ws.Close()` in here based on the name.